pkg/crdschema/gen.go

```diff
@@ -0,0 +1,377 @@
+	ManifestsPath string `marker:"manifests,optional"`
+	// If true, no file is written. The Generate func will return an
+	// error if a file on disk differs from the generated manifest.
+	VerifyOnly bool `marker:",optional"`
+}
+
+var _ genall.Generator = &Generator{}
+
+func (Generator) RegisterMarkers(into *markers.Registry) error {
+	return crdmarkers.Register(into)
+}
+
+func (g Generator) Generate(ctx *genall.GenerationContext) (result error) {
+	p := &crdgen.Parser{
+		Collector: ctx.Collector,
+		Checker:   ctx.Checker,
+	}
+	f := crdgen.Flattener{
+		Parser: p,
+	}
+
+	crdgen.AddKnownTypes(p)
+	for _, root := range ctx.Roots {
+		p.NeedPackage(root)
+	}
+
+	metav1Pkg := crdgen.FindMetav1(ctx.Roots)
+	if metav1Pkg == nil {
+		// no objects in the roots, since nothing imported metav1
+		fmt.Printf("No metav1 import found in any of the API packages.\n")
+		return nil
+	}
+
+	// load existing manifests from manifests/ dir
+	existingCRDsByFileNames, err := crdsFromDirectory(ctx, g.ManifestsPath)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("Found %d CRD manifests.\n", len(existingCRDsByFileNames))
+	existingFileNamesByGroupKind := map[schema.GroupKind]string{}
+	for fn, crd := range existingCRDsByFileNames {
+		existingFileNamesByGroupKind[crd.GroupKind] = fn
+	}
+
+	changedFiles := []string{}
+	for _, groupKind := range crdgen.FindKubeKinds(p, metav1Pkg) {
+		fn, found := existingFileNamesByGroupKind[groupKind]
+		if !found {
+			continue
+		}
+		existing := existingCRDsByFileNames[fn]
+
+		// get existing versions
+		existingVersions := make(map[string]bool, 0)
+		if existing.CRD.Spec.Version != "" {
+			existingVersions[existing.CRD.Spec.Version] = true
+		}
+		for _, v := range existing.CRD.Spec.Versions {
+			existingVersions[v.Name] = true
+		}
+
+		// go through source code versions
+		versionSchemas := map[string]*apiextensionsv1beta1.JSONSchemaProps{}
+		for pkg, gv := range p.GroupVersions {
+			if gv.Group != groupKind.Group {
+				continue
+			}
+			if !existingVersions[gv.Version] {
+				fmt.Printf("Source code for %s in %s/%s not found.\n", groupKind.Kind, gv.Group, gv.Version)
+				continue
+			}
+			versionSchemas[gv.Version] = nil
+
+			// find parser schema
+			typeIdent := crdgen.TypeIdent{Package: pkg, Name: groupKind.Kind}
+			typeInfo := p.Types[typeIdent]
+			if typeInfo == nil {
+				continue
+			}
+
+			// generate flattened schema
+			p.NeedSchemaFor(typeIdent)
+			versionSchemas[gv.Version] = crdgen.FlattenEmbedded(f.FlattenType(typeIdent))
+		}
+
+		// edge cases: no schema or all nil
+		allNil := true
+		for _, s := range versionSchemas {
+			if s != nil {
+				allNil = false
+				break
+			}
+		}
+		if allNil {
+			fmt.Printf("No schema found for %s in %s, skipping.\n", groupKind.Kind, groupKind.Group)
+			continue
+		}
+
+		// use global schema if there is just one, or all are equal
+		allEqual := true
+		var first *apiextensionsv1beta1.JSONSchemaProps
+		for _, s := range versionSchemas {
```

Comment:
I suggest to change this for loop to the following
```
if len(versionSchemas) > 0 {
  first = versionSchemas[0]
}
```

Reply:
done